src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -143,20 +101,25 @@ def run_python_test(test_target, pytest, python_setup, source_root_config, pex_b
   all_input_digests = [
     sources_digest,
     inits_digest.directory_digest,
-    requirements_pex_response.output_directory_digest,
+    resolved_requirements_pex.directory_digest,
   ]
   merged_input_files = yield Get(
     Digest,
     DirectoriesToMerge,
     DirectoriesToMerge(directories=tuple(all_input_digests)),
   )
 
+  interpreter_search_paths = text_type(create_path_env_var(python_setup.interpreter_search_paths))
   pex_exe_env = {'PATH': interpreter_search_paths}
   # TODO(#6071): merge the two dicts via ** unpacking once we drop Py2.
   pex_exe_env.update(subprocess_encoding_environment.invocation_environment_dict)
 
+  # NB: we use the hardcoded and generic bin name `python`, rather than something dynamic like
+  # `sys.executable`, to ensure that the interpreter may be discovered both locally and in remote
+  # execution. This is only used to run the downloaded PEX tool; it is not necessarily the
+  # interpreter that PEX will use to execute the generated .pex file.
   request = ExecuteProcessRequest(
-    argv=(python_binary, './{}'.format(output_pytest_requirements_pex_filename)),
+    argv=("python", './{}'.format(output_pytest_requirements_pex_filename)),
```

Comment:
I am not entirely sure this is kosher; I think it happens to work locally because of how we happen to have implemented `$PATH` suppression (https://github.com/pantsbuild/pants/blob/6ab5f7aed81dc623e0edcebd60af1f9d28fc98c5/src/rust/engine/process_execution/src/local.rs#L106-L114), but according to the remote execution API `$PATH` shouldn't be used to look up `argv[0]` even if it's set.

Sadly I suspect the technically correct answer here is a little wrapper shell script along the lines of:

`exec "$@"`

which we'd invoke `lookup_on_path python ./output...`

but I guess we can punt on this until it becomes a problem.

(We could also make our local process execution resemble the spec more closely by before running inspecting "does `argv[0]` contain a `/`, if not, error"?)

Reply:
This works on remoting and local. I made this change in https://github.com/pantsbuild/pants/pull/7844 specifically to fix the remoting case.

The reason it works (I think) is that we explicitly set the `env` to include the entry `PATH: interpreter_search_paths`.

--

Are you proposing having this be something like `/usr/bin/python`? Or parametrizing it as an option so that you can set which you want?

Reply:
As per https://github.com/bazelbuild/remote-apis/blob/a5c577357528b33a4adff88c0c7911dd086c6923/build/bazel/remote/execution/v2/remote_execution.proto#L433-L436 the execution system _shouldn't_ do $PATH lookup; in particular, it's totally valid for it to `execv` and not `execvp`... Our local execution happens to `execvp` because that's what rust conveniently exposes to us, but it shouldn't...

I'm surprised RBE does, if it does... Does it?

Reply:
> I'm surprised RBE does, if it does... Does it?

I haven't worked on remoting in two weeks and the context switch to verify this is going to take too much time to be able to quickly check, but I know when working on https://github.com/pantsbuild/pants/pull/7844 the change allowed me to get further than before. Originally, iirc it would complain it could not find the file `/Users/eric/DocsLocal/code/projects/pants/build-support/pants_dev_deps.py36.venv/bin/python`; with the change, the error was just that Pex could not resolve requirements.

Will hopefully be done with Py3 work soon and be able to go back to remoting!